tests/test_boundaries.py

```diff
@@ -40,53 +39,83 @@ def test_equilibrium_boundary_pu(f_lattice):
     mask = (f[0] > 0).cpu().numpy()  # will contain all points
     units = UnitConversion(lattice, reynolds_number=1)
     pressure = 0
-    velocity = 0.1*np.ones(lattice.D)
+    velocity = 0.1 * np.ones(lattice.D)
     feq = lattice.equilibrium(
-            lattice.convert_to_tensor(units.convert_pressure_pu_to_density_lu(pressure)),
-            lattice.convert_to_tensor(units.convert_velocity_to_lu(velocity))
-        )
+        lattice.convert_to_tensor(units.convert_pressure_pu_to_density_lu(pressure)),
+        lattice.convert_to_tensor(units.convert_velocity_to_lu(velocity))
+    )
     feq_field = torch.einsum("q,q...->q...", feq, torch.ones_like(f))
 
     eq_boundary = EquilibriumBoundaryPU(mask, lattice, units, velocity=velocity, pressure=pressure)
     f = eq_boundary(f)
 
     assert f.cpu().numpy() == pytest.approx(feq_field.cpu().numpy())
-    #assert f.cpu().numpy() == pytest.approx(f_old.cpu().numpy())
+    # assert f.cpu().numpy() == pytest.approx(f_old.cpu().numpy())
 
 
 def test_anti_bounce_back_outlet(f_lattice):
     """Compares the result of the application of the boundary to f to the result using the formula taken from page 195
     of "The lattice Boltzmann method" (2016 by Krüger et al.) if both are similar it is assumed to be working fine."""
     f, lattice = f_lattice
-    #generates reference value of f using non-dynamic formula
+    # generates reference value of f using non-dynamic formula
     f_ref = f
     u = lattice.u(f)
     D = lattice.stencil.D()
     Q = lattice.stencil.Q()
+
     if D == 3:
         direction = [1, 0, 0]
+
         if Q == 27:
             u_w = u[:, -1, :, :] + 0.5 * (u[:, -1, :, :] - u[:, -2, :, :])
+            u_w_norm = torch.norm(u_w, dim=0)
+
             for i in [1, 11, 13, 15, 17, 19, 21, 23, 25]:
-                f_ref[lattice.stencil.opposite[i], -1, :, :] = - f_ref[i, -1, :, :] + lattice.stencil.w[i] * lattice.rho(f)[0, -1, :, :] * \
-                    (2 + torch.einsum('c, cyz -> yz', torch.tensor(lattice.stencil.e[i], device=f.device, dtype=f.dtype), u_w) ** 2 / lattice.stencil.cs ** 4 - (torch.norm(u_w, dim=0) / lattice.stencil.cs) ** 2)
+                stencil_e_tensor = torch.tensor(lattice.stencil.e[i], device=f.device, dtype=f.dtype)
```

Comment:
I think this format is acceptable.
Can someone confirm that I have not done any typos in this file?

Reply:
lgtm